crates/hir_ty/src/infer/expr.rs

```diff
@@ -137,24 +137,28 @@ impl<'a> InferenceContext<'a> {
 
                 self.coerce_merge_branch(&then_ty, &else_ty)
             }
-            Expr::Block { statements, tail, label, id: _ } => match label {
-                Some(_) => {
-                    let break_ty = self.table.new_type_var();
-                    self.breakables.push(BreakableContext {
-                        may_break: false,
-                        break_ty: break_ty.clone(),
-                        label: label.map(|label| self.body[label].name.clone()),
-                    });
-                    let ty = self.infer_block(statements, *tail, &Expectation::has_type(break_ty));
-                    let ctxt = self.breakables.pop().expect("breakable stack broken");
-                    if ctxt.may_break {
-                        ctxt.break_ty
-                    } else {
-                        ty
+            Expr::Block { statements, tail, label, id: _ } => {
+                self.resolver = resolver_for_expr(self.db.upcast(), self.owner, tgt_expr);
```

Comment:
Can we push directly a new scope onto the `Resolver`, then pop it at the end of the block, so we reuse it?

This feels a bit iffy -- we replace the old one, but keep using the new even after the end of the block.

Reply:
Oh, yeah, I meant to swap the old one back

Reply:
> Can we push directly a new scope onto the `Resolver`, then pop it at the end of the block, so we reuse it?

We can do that, but that sort of feels like mixing concerns. I think type inference shouldn't really know about name resolution scoping rules.

Perhaps a nicer solution would be to have an HIR body visitor that gives you the correct `Resolver` for every expression?

Reply:
Maybe. Or at least an API like `self.resolver = resolver_for_expr(self.db.upcast(), self.owner, tgt_expr, prev_resolver); /* ... */ self.resolver.pop();`?